tests/core/arch-arm64/arm64stalker-fixture.c

```diff
@@ -0,0 +1,203 @@
+    (gum_fake_event_sink_get_nth_event_as_exec (fixture->sink, N)->location)
+
+typedef struct _TestArm64StalkerFixture
+{
+  GumStalker * stalker;
+  GumFakeEventSink * sink;
+
+  guint8 * code;
+  guint8 * last_invoke_calladdr;
+  guint8 * last_invoke_retaddr;
+} TestArm64StalkerFixture;
+
+typedef gint (* StalkerTestFunc) (gint arg);
+
+static void silence_warnings (void);
+
+static void
+debug_hello (gpointer pointer)
+{
+  g_print ("* pointer: %p *\n", pointer);
+}
+
+static void
+put_debug_print (GumArm64Writer * cw,
+                 gpointer pointer)
+{
+  gum_arm64_writer_put_push_all_registers (cw);
+  gum_arm64_writer_put_call_address_with_arguments (cw,
+      GUM_FUNCPTR_TO_POINTER (debug_hello), 1,
+      GUM_ARG_ADDRESS, GUM_ADDRESS (pointer));
+  gum_arm64_writer_put_pop_all_registers (cw);
+}
+
+static void
+put_debug_print_reg (GumArm64Writer * cw,
+                     arm64_reg reg)
+{
+  gum_arm64_writer_put_push_all_registers (cw);
+  gum_arm64_writer_put_call_address_with_arguments (cw,
+      GUM_FUNCPTR_TO_POINTER (debug_hello), 1,
+      GUM_ARG_REGISTER, reg);
+  gum_arm64_writer_put_pop_all_registers (cw);
+}
+
+static void
+test_arm64_stalker_fixture_setup (TestArm64StalkerFixture * fixture,
+                                  gconstpointer data)
+{
+  fixture->stalker = gum_stalker_new ();
+  fixture->sink = GUM_FAKE_EVENT_SINK (gum_fake_event_sink_new ());
+
+  silence_warnings ();
+}
+
+static void
+test_arm64_stalker_fixture_teardown (TestArm64StalkerFixture * fixture,
+                                     gconstpointer data)
+{
+  g_object_unref (fixture->sink);
+  g_object_unref (fixture->stalker);
+
+  if (fixture->code != NULL){
+    gum_free_pages (fixture->code);
+  }
+}
+
+static guint8 *
+test_arm64_stalker_fixture_dup_code (TestArm64StalkerFixture * fixture,
+                                     const guint32 * tpl_code,
+                                     guint tpl_size)
+{
+  GumAddressSpec spec;
+
+  spec.near_address = gum_stalker_follow_me;
+  spec.max_distance = G_MAXINT32 / 2;
+
+  if (fixture->code != NULL){
+    gum_free_pages (fixture->code);
+  }
+  fixture->code = (guint32 *) gum_alloc_n_pages_near (
+      (tpl_size / gum_query_page_size ()) + 1, GUM_PAGE_RWX, &spec);
+  memcpy (fixture->code, tpl_code, tpl_size);
+  return fixture->code;
+}
+
+# define INVOKER_INSN_COUNT 6
+# define INVOKER_IMPL_OFFSET 2
+
+/* custom invoke code as we want to stalk a deterministic code sequence */
+static gint
+test_arm64_stalker_fixture_follow_and_invoke (TestArm64StalkerFixture * fixture,
+                                              StalkerTestFunc func,
+                                              gint arg)
+{
+  GumAddressSpec spec;
+  gint ret;
+  guint8 * code;
+  GumArm64Writer cw;
+  GCallback invoke_func;
+
+  spec.near_address = gum_stalker_follow_me;
+  spec.max_distance = G_MAXINT32 / 2;
+  code = (guint8 *) gum_alloc_n_pages_near (1, GUM_PAGE_RWX, &spec);
+
+  // init writer
```

Comment:
Also, let's avoid comments like this:
```
  // init writer
  gum_arm64_writer_init (&cw, code);
```
^ the code already conveys the same thing that the comment is saying.